### PR TITLE
Fix Impossible Mission II transition effect

### LINKED ISSUE #3447: Impossible Mission II: Visual Glitch on Tandy
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

_No response_

### What Operating System are you using?

Linux x86_64

### If Other OS, please describe

Arch Linux

### Relevant hardware info

_No response_

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

As the first level is loading in, there is an animation where the screen starts out as a narrow band in the center and then expands out vertically in both directions until it fills the screen completely.

This animation looks correct in v0.80.1 and incorrect on main.  This is a regression that I have bisected to commit dd8ae1423a9d496f78fdcf39798568aea6bbadca

Note there is a (mostly) unrelated divide by zero crash that must be worked around if checking out older commits.

### Steps to reproduce the behaviour.

Explain how to reproduce
1.  Put `machine = tandy` in your config
2. Run IM.COM or IM2_TAN .EXE (the former seems to auto-detect machine type and load the latter)
3. Skip past intro and "Load save game" screen (just press enter)
4. Observe the animation as the level loads in

### Download URL of affected game or software

https://www.abandonwaredos.com/abandonware-game.php?abandonware=Impossible+Mission+2&gid=1873

### Your configuration

```ini
[dosbox]
machine = tandy
```

### Provide a Log

_No response_

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

### PR DESCRIPTION
**Pull request by weirddan455**
# Description

First commit addresses a regression caused by b68b66f07831c91f197acf0b06f847c28c9a7946.  It causes flickering when the video mode changes quickly during the transition effect.

Second commit addresses a regression caused by https://github.com/dosbox-staging/dosbox-staging/commit/dd8ae1423a9d496f78fdcf39798568aea6bbadca.  Previously, the aspect ratio for Tandy was hardcoded to 5/6.  We have since added some complexity where pixel-doubling and double-scanning need to be taken into account.  However, the width needs to be hard-coded to 200 so this effect renders correctly.

## Related issues

Fixes #3447

# Manual testing

With `machine = tandy` tested the following with both `glshader=crt-auto` (pixel doubling enabled) and `glshader=sharp` (pixel doubling disabled).

- Tested Impossible Mission II (160 x 200)
- Tested King's Quest 4 (320 x 200)
- Tested Space Quest 1 (TANDY 320 x 200 plus CGA text-mode 320x200)

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

File: src/hardware/vga_other.cpp [hunk 1]
```diff
@@ -92,8 +92,11 @@ static void write_crtc_data_other(io_port_t, io_val_t value, io_width_t)
 	case 0x06:		//Vertical rows
 		// Impossible Mission II sets this to zero briefly
 		// This leads to a divide by zero crash if VGA resize code is run
-		if ((val > 0) && (vga.other.vdend ^ val)) VGA_StartResize();
-		vga.other.vdend = val;
+		if (val > 0 && val != vga.other.vdend) {
+			vga.other.vdend = val;
+			// The default half-frame peroid delay leads to flickering
```
Comment (johnnovak) [inline src/hardware/vga_other.cpp:97]: typo: period

Comment (johnnovak) [reply src/hardware/vga_other.cpp:97]: "leads to flickering in the level start effect of Impossible Mission II on Tandy"

File: src/hardware/vga_draw.cpp [hunk 2]
```diff
@@ -2054,6 +2054,13 @@ ImageInfo setup_drawing()
 		return pcjr_or_tga();
 	};
 
+	// As far as I can tell, all Tandy modes have a height of 200.
```
Comment (johnnovak) [inline src/hardware/vga_draw.cpp:2057]: Let's be more assertive about this 😄 "All Tandy modes have a height of 200 pixels."

File: src/hardware/vga_draw.cpp [hunk 3]
```diff
@@ -2054,6 +2054,13 @@ ImageInfo setup_drawing()
...
+	// Some games (ex. Impossible Mission II) fiddle with vga.other.vdend
+	// The result of this should be rendering a short (by height) image in the horizonal center
+	// With black on the top/bottom
```
Comment (johnnovak) [inline src/hardware/vga_draw.cpp:2060]: minor: join the above two lines into a single sentence, period at the end.

**Comment by johnnovak** [comment]
Very nice @weirddan455 🎉 

We should apply the fix to the composite Tandy modes as well. If you run the game with this config, the problem is back:

```ini
[dosbox]
machine = tandy

[composite]
composite = on
```

The solution is to apply the fix to these paths as well:

```
M_CGA4_COMPOSITE
M_CGA2_COMPOSITE
```

File: src/hardware/vga_draw.cpp [hunk 4]
```diff
@@ -2054,6 +2054,13 @@ ImageInfo setup_drawing()
...
+	// Use this hard-coded value when calculating pixel aspect ratio so this effect looks correct.
+	constexpr uint16_t TandyAspectHeight = 200;
```
Comment (johnnovak) [inline src/hardware/vga_draw.cpp:2062]: Given we need to fix the composite paths too, and those composite modes affect CGA too, maybe call this `CgaTandyAspectHeight`? Well, PCjr would be affected too, but then the name would be even sillier...

File: src/hardware/vga_other.cpp (hunk 1 above)
Comment (weirddan455) [reply src/hardware/vga_other.cpp]: fixed

File: src/hardware/vga_draw.cpp (hunk 2 above)
Comment (weirddan455) [reply src/hardware/vga_draw.cpp]: fixed

File: src/hardware/vga_draw.cpp (hunk 3 above)
Comment (weirddan455) [reply src/hardware/vga_draw.cpp]: fixed

File: src/hardware/vga_draw.cpp (hunk 4 above)
Comment (weirddan455) [reply src/hardware/vga_draw.cpp]: Yep, good catch.  Applied the change to the composite modes and confirmed it looks correct there as well.

**Review by johnnovak — APPROVED** [review]
Nice job @weirddan455 , ship it 😎

**Pull request merged**